File: src/services/ragService.ts
/**
 * Complete Retrieval-Augmented Generation (RAG) Service
 * ====================================================
 * Connects Qdrant vector retrieval with OpenRouter LLM answer synthesis.
 * Implements strict context grounding, citations mapping, and fallback handling.
 */

import { defaultRetrieverService, RetrievedChunk, RetrievalResult } from './retrieverService';
import { defaultLLMService, LLMService } from './llmService';

export interface CitationItem {
  document_name: string;
  page_number: number;
  retrieved_text: string;
}

export interface RAGPipelineResult {
  status: 'success' | 'error';
  answer: string;
  citations: CitationItem[];
  retrieval_result: RetrievalResult;
  logs: string[];
  error_message?: string;
}

export const NO_INFO_FALLBACK = "The information is not available in the provided documents.";

export class RAGService {
  private topK: number;
  private similarityThreshold: number;
  private maxContextChars: number;
  private llmService: LLMService;

  constructor(
    topK: number = 5,
    similarityThreshold: number = 0.05,
    maxContextChars: number = 3500,
    llmService: LLMService = defaultLLMService
  ) {
    this.topK = topK;
    this.similarityThreshold = similarityThreshold;
    this.maxContextChars = maxContextChars;
    this.llmService = llmService;
  }

  private normalizeText(text: string): string {
    return text.toLowerCase().replace(/\s+/g, ' ').trim();
  }

  /**
   * Executes the complete RAG Pipeline.
   */
  public async executePipeline(question: string): Promise<RAGPipelineResult> {
    const logs: string[] = [];

    const addLog = (msg: string) => {
      logs.push(`[${new Date().toISOString()}] ${msg}`);
      console.log(`[RAGService] ${msg}`);
    };

    const cleanedQuestion = question.trim();

    // 1. Validate empty question
    if (!cleanedQuestion) {
      addLog('Empty question provided.');
      return {
        status: 'error',
        answer: NO_INFO_FALLBACK,
        citations: [],
        retrieval_result: {
          status: 'error',
          query: '',
          retrieved_documents_count: 0,
          retrieved_pages: [],
          similarity_scores: [],
          chunks: [],
          logs: [],
          error_message: 'Question cannot be empty.',
        },
        logs,
        error_message: 'Please enter a question.',
      };
    }

    // 2. Perform Semantic Vector Retrieval from Qdrant 'pdf_documents' collection
    addLog('Executing vector retrieval from Qdrant...');
    const retrievalResult = await defaultRetrieverService.retrieve(cleanedQuestion, this.topK);

    // Append retriever internal logs
    for (const rlog of retrievalResult.logs) {
      addLog(rlog.message);
    }

    // Check raw chunks retrieved
    const rawChunks: RetrievedChunk[] = retrievalResult.chunks || [];
    if (rawChunks.length === 0) {
      addLog('No relevant chunks found in Qdrant vector store. Skipping LLM call.');
      return {
        status: 'success',
        answer: NO_INFO_FALLBACK,
        citations: [],
        retrieval_result: retrievalResult,
        logs,
      };
    }

    // 3. Log similarity scores & filter by Similarity Threshold
    addLog(`Retrieved ${rawChunks.length} raw candidate chunks.`);
    rawChunks.forEach((c, idx) => {
      addLog(`Chunk #${idx + 1} score: ${c.similarity_score.toFixed(4)} (Doc: ${c.document_name}, Page: ${c.page_number})`);
    });

    const filteredChunks = rawChunks.filter(
      (c) => c.similarity_score >= this.similarityThreshold
    );

    const discardedByScore = rawChunks.length - filteredChunks.length;
    if (discardedByScore > 0) {
      addLog(`Discarded ${discardedByScore} chunk(s) falling below similarity threshold (${this.similarityThreshold}).`);
    }

    if (filteredChunks.length === 0) {
      addLog(`No chunks satisfied the similarity threshold (${this.similarityThreshold}). Returning fallback.`);
      return {
        status: 'success',
        answer: NO_INFO_FALLBACK,
        citations: [],
        retrieval_result: retrievalResult,
        logs,
      };
    }

    // 4. Remove duplicate or near-identical text chunks
    const seenTexts = new Set<string>();
    const dedupedChunks: RetrievedChunk[] = [];

    for (const c of filteredChunks) {
      const norm = this.normalizeText(c.text);
      if (!seenTexts.has(norm)) {
        seenTexts.add(norm);
        dedupedChunks.push(c);
      } else {
        addLog(`Discarded duplicate chunk text from Doc: ${c.document_name}, Page: ${c.page_number}.`);
      }
    }

    addLog(`Selected ${dedupedChunks.length} unique chunk(s) for context construction.`);

    // 5. Context Optimization (character budget capping)
    const contextBlocks: string[] = [];
    const citationsMap = new Map<string, CitationItem>();
    let currentChars = 0;

    for (let i = 0; i < dedupedChunks.length; i++) {
      const chunk = dedupedChunks[i];
      const docName = chunk.document_name || 'document.pdf';
      const pageNum = chunk.page_number || 1;
      const text = chunk.text.trim();

      const block = `[Chunk ${i + 1} | Document: ${docName} | Page: ${pageNum}]\n${text}`;
      if (currentChars + block.length > this.maxContextChars && i > 0) {
        addLog(`Reached max context character limit (${this.maxContextChars} chars). Capping context at ${i} chunks.`);
        break;
      }

      contextBlocks.push(block);
      currentChars += block.length;

      const citationKey = `${docName}-p${pageNum}-${text.slice(0, 40)}`;
      if (!citationsMap.has(citationKey)) {
        citationsMap.set(citationKey, {
          document_name: docName,
          page_number: pageNum,
          retrieved_text: text,
        });
      }
    }

    const combinedContext = contextBlocks.join('\n\n');
    const userPrompt = `Context:\n${combinedContext}\n\nQuestion:\n${cleanedQuestion}\n\nAnswer:`;

    addLog('Generating prompt...');
    console.log('=== EXACT GENERATED RAG USER PROMPT ===');
    console.log(userPrompt);
    console.log('======================================');
    addLog(`Sending LLM request with context length: ${combinedContext.length} chars.`);
    addLog(`Exact Prompt:\n${userPrompt}`);

    // 6. Call OpenRouter LLM Service
    addLog('Calling OpenRouter...');
    addLog('Generating response...');

    try {
      const generatedAnswer = await this.llmService.generateCompletion(userPrompt);
      addLog('Response completed.');

      if (!generatedAnswer || generatedAnswer.includes(NO_INFO_FALLBACK)) {
        addLog('LLM indicated information not available in context.');
        return {
          status: 'success',
          answer: NO_INFO_FALLBACK,
          citations: [],
          retrieval_result: retrievalResult,
          logs,
        };
      }

      const finalCitations = Array.from(citationsMap.values());

      return {
        status: 'success',
        answer: generatedAnswer,
        citations: finalCitations,
        retrieval_result: retrievalResult,
        logs,
      };
    } catch (err: any) {
      addLog(`LLM invocation error: ${err?.message || err}`);
      return {
        status: 'error',
        answer: NO_INFO_FALLBACK,
        citations: [],
        retrieval_result: retrievalResult,
        logs,
        error_message: `LLM service error: ${err?.message || err}`,
      };
    }
  }
}

export const defaultRAGService = new RAGService(5, 0.05, 3500);

File: rag_service.py
"""
Complete Retrieval-Augmented Generation (RAG) Service Module
=============================================================

This module orchestrates the complete RAG pipeline:
1. Performs semantic vector retrieval from Qdrant (`pdf_documents` collection).
2. Validates retrieved context (returns non-hallucination fallback if context is empty/irrelevant without calling LLM).
3. Constructs formatted prompt with context ONLY.
4. Invokes OpenRouter LLM service under strict system rules.
5. Formats structured answer with citations.

Response Format:
{
    "answer": "<Generated Answer>",
    "citations": [
        {
            "document_name": "<name>",
            "page_number": <int>,
            "retrieved_text": "<text>"
        }
    ]
}

Logging Sequence:
- Generating prompt...
- Calling OpenRouter...
- Generating response...
- Response completed.
"""

import logging
from typing import List, Dict, Any, Optional

from retriever import SemanticRetriever, retrieve_relevant_chunks
from llm_service import OpenRouterLLMService, SYSTEM_PROMPT

logger = logging.getLogger("RAGService")
logging.basicConfig(
    level=logging.INFO,
    format="[%(asctime)s] [%(levelname)s] [%(name)s] %(message)s"
)

NO_INFO_FALLBACK = "The information is not available in the provided documents."


class RAGPipelineService:
    """
    Production RAG Pipeline Service linking semantic vector retrieval with LLM answer synthesis.
    Includes similarity score thresholding, duplicate removal, context length optimization, and detailed logging.
    """

    def __init__(
        self,
        retriever: Optional[SemanticRetriever] = None,
        llm_service: Optional[OpenRouterLLMService] = None,
        top_k: int = 5,
        similarity_threshold: float = 0.15,
        max_context_chars: int = 3500
    ) -> None:
        """
        Initialize RAG Pipeline Service.

        Args:
            retriever (SemanticRetriever, optional): Vector retriever instance.
            llm_service (OpenRouterLLMService, optional): OpenRouter LLM service instance.
            top_k (int): Number of top chunks to retrieve for prompt context (default: 5).
            similarity_threshold (float): Minimum cosine similarity score required (default: 0.15).
            max_context_chars (int): Maximum combined context length limit (default: 3500).
        """
        self.top_k = top_k
        self.similarity_threshold = similarity_threshold
        self.max_context_chars = max_context_chars
        self.retriever = retriever or SemanticRetriever(top_k=top_k)
        self.llm_service = llm_service or OpenRouterLLMService()

    @staticmethod
    def _normalize_text(text: str) -> str:
        """Normalize text for duplicate chunk detection."""
        return " ".join(text.lower().split())

    def run_pipeline(self, question: str) -> Dict[str, Any]:
        """
        Execute full RAG pipeline for a given user question.

        Args:
            question (str): User question text.

        Returns:
            Dict[str, Any]: Result containing answer string and citations list.
        """
        if not question or not question.strip():
            logger.warning("Empty question submitted to RAG pipeline.")
            return {
                "answer": NO_INFO_FALLBACK,
                "citations": []
            }

        cleaned_question = question.strip()

        # Step 1: Retrieve Top 5 relevant chunks from Qdrant
        retrieval_res = self.retriever.retrieve(cleaned_question, top_k=self.top_k)
        raw_chunks = retrieval_res.get("results", [])

        if not raw_chunks:
            logger.info("No relevant chunks found in Qdrant vector store. Skipping LLM call.")
            return {
                "answer": NO_INFO_FALLBACK,
                "citations": []
            }

        # Step 2: Log similarity scores & filter by Similarity Threshold
        logger.info(f"Retrieved {len(raw_chunks)} raw candidate chunks.")
        for idx, c in enumerate(raw_chunks, 1):
            logger.info(f"Chunk #{idx} similarity score: {c.get('similarity_score', 0.0):.4f} (Doc: {c.get('document_name')}, Page: {c.get('page_number')})")

        filtered_chunks = [
            c for c in raw_chunks
            if c.get("similarity_score", 0.0) >= self.similarity_threshold
        ]

        discarded_by_threshold = len(raw_chunks) - len(filtered_chunks)
        if discarded_by_threshold > 0:
            logger.info(f"Discarded {discarded_by_threshold} chunk(s) falling below similarity threshold ({self.similarity_threshold}).")

        if not filtered_chunks:
            logger.info(f"No chunks satisfied the similarity threshold ({self.similarity_threshold}). Returning fallback.")
            return {
                "answer": NO_INFO_FALLBACK,
                "citations": []
            }

        # Step 3: Remove duplicate or near-identical chunks
        seen_texts = set()
        deduped_chunks = []
        for c in filtered_chunks:
            norm = self._normalize_text(c.get("text", ""))
            if norm not in seen_texts:
                seen_texts.add(norm)
                deduped_chunks.append(c)
            else:
                logger.info(f"Discarded duplicate chunk text from Doc: {c.get('document_name')}, Page: {c.get('page_number')}.")

        logger.info(f"Selected {len(deduped_chunks)} unique chunk(s) for context construction.")

        # Step 4: Context Optimization (character budget capping)
        context_blocks = []
        citations = []
        current_chars = 0

        for idx, chunk in enumerate(deduped_chunks, 1):
            doc_name = chunk.get("document_name", "document.pdf")
            page_num = chunk.get("page_number", 1)
            chunk_text = chunk.get("text", "").strip()

            block = f"[Chunk {idx} | Document: {doc_name} | Page: {page_num}]\n{chunk_text}"
            if current_chars + len(block) > self.max_context_chars and idx > 1:
                logger.info(f"Reached max context character limit ({self.max_context_chars} chars). Capping context at {idx - 1} chunks.")
                break

            context_blocks.append(block)
            current_chars += len(block)

            citations.append({
                "document_name": doc_name,
                "page_number": page_num,
                "retrieved_text": chunk_text
            })

        combined_context = "\n\n".join(context_blocks)

        logger.info("Generating prompt...")
        user_prompt = (
            f"Context:\n{combined_context}\n\n"
            f"Question:\n{cleaned_question}\n\n"
            "Answer:"
        )

        logger.info(f"=== EXACT GENERATED RAG USER PROMPT ===\n{user_prompt}\n=======================================")
        logger.info(f"Sending LLM request with context length: {len(combined_context)} chars.")

        # Step 5: Call LLM
        generated_answer = self.llm_service.generate_completion(
            user_prompt=user_prompt,
            system_prompt=SYSTEM_PROMPT
        )

        logger.info("LLM response received.")

        if NO_INFO_FALLBACK in generated_answer or not generated_answer.strip():
            logger.info("LLM indicated information not available in context.")
            return {
                "answer": NO_INFO_FALLBACK,
                "citations": []
            }

        return {
            "answer": generated_answer,
            "citations": citations
        }


# Convenience module function
def answer_question_with_rag(question: str) -> Dict[str, Any]:
    """
    Convenience function to run complete RAG pipeline.

    Usage:
        res = answer_question_with_rag("What is vector retrieval?")
        print(res["answer"])
        print("Citations:", res["citations"])
    """
    pipeline = RAGPipelineService()
    return pipeline.run_pipeline(question)


if __name__ == "__main__":
    test_q = "Explain retrieval augmented generation."
    output = answer_question_with_rag(test_q)
    print("RAG Pipeline Output:")
    print("Answer:", output["answer"])
    print("Citations Count:", len(output["citations"]))
